_posts/2019-02-15-snapshots-of-buggy-black-boxes.md

```diff
@@ -0,0 +1,128 @@
+---
+layout: post
+title: Snapshots of Buggy Black Boxes
+author: mpajunen
+excerpt: Snapshot testing has been gaining popularity in frontend development. Could it be useful for testing legacy code as well?
+tags:
+- legacy code
+- testing
+---
+
+Snapshot tests have recently become popular in testing UI components. We decided to try snapshots in testing legacy code.
+
+## A Black Box of Code
+
+A while ago we were faced with a typical situation when working with legacy code: A largely unknown and practically untested Black Box. This Black Box had been lifted from an older system and rewritten for the new platform already. Unfortunately, much of the old cruft remained.
+
+We already had a list of new features and improvements to implement. Looking at the code this seemed like a tall order: Data fetches, business logic and output marshalling were all gloriously mixed up. It didn't take long to realize that significant refactoring was needed before getting to the new features.
+
+### Testing Black Boxes
+
+Faced with the refactoring task, we wanted to make sure not to break too many things that were already working. This meant we'd need to write at least some tests first.
+
+A major problem with untested code is that it's often untestable as well. If the code blocks are long and deeply nested and functions are even longer or common state is modified without care, writing meaningful tests can be impossible.
+
+Writing unit tests for the existing code was quickly ruled out except for a few isolated sections. Integration tests proved more reasonable. Simply initializing the relevant state with known values and matching the output increased our confidence to make changes quite a bit. Some bugs were also discovered right away.
+
+Handling these bugs turned out to be the next challenge. Unit tests had proved nearly impossible and integration tests looked like a lot of work: The lengthy messages in the assertions would have to be updated after any intentional changes or fixes.
+
+Fortunately, we already had a solution in mind. A bit earlier we had started using snapshot testing for frontend components. The Black Box seemed like a similar if slightly more complex problem: Set some initial values, get a result, check that it matches a known value.
```

Comment:
Fortunately, we already thought a solution for this problem?

Reply:
Jättäisin A bit pois tuosta seuraavasta lauseesta.

Reply:
"A bit earlier" ---> "We had recently started using snapshot testing...."

Reply:
Vaihdoin ensimmäisen: "Fortunately, we had already thought of a solution."